go-server-server/main.go

```diff
@@ -14,74 +14,162 @@ import (
-        // RequireAndVerifyClientCert
-        ClientAuth: tls.RequireAndVerifyClientCert,
-        MinVersion: tls.VersionTLS12,
-    }
-
-    tlsConfig.BuildNameToCertificate()
-
-    server := &http.Server{
-        Addr:      ":8081",
-        Handler:   handler,
-        TLSConfig: tlsConfig,
+func StartHttpsServer(handler http.Handler, messenger <-chan int, wgroup *sync.WaitGroup) {
+    defer wgroup.Done()
+    for {
+        clientCert, err := ioutil.ReadFile(*sw.ClientCertFlag)
+        if err != nil {
+            log.Fatalf("error: couldn't open client cert file, %s", err)
+        }
+        clientCertPool := x509.NewCertPool()
+        clientCertPool.AppendCertsFromPEM(clientCert)
+
+        // Setup HTTPS client cert the server trust and validation policy
+        tlsConfig := &tls.Config{
+            ClientCAs: clientCertPool,
+            // NoClientCert
+            // RequestClientCert
+            // RequireAnyClientCert
+            // VerifyClientCertIfGiven
+            // RequireAndVerifyClientCert
+            ClientAuth: tls.RequireAndVerifyClientCert,
+            MinVersion: tls.VersionTLS12,
+        }
+        tlsConfig.BuildNameToCertificate()
+
+        server := &http.Server{
+            Addr:      ":8081",
+            Handler:   handler,
+            TLSConfig: tlsConfig,
+        }
+
+        log.Printf("info: https endpoint started")
+
+        // Listening should happen in a go-routine to prevent blocking
+        go func() {
+            if err := server.ListenAndServeTLS(*sw.ServerCertFlag, *sw.ServerKeyFlag); err != nil {
+                log.Println(err)
+            }
+        }()
+
+        value := <-messenger
+        log.Printf("info: HTTPS Signal received. Shutting down...")
+        if err := server.Shutdown(context.Background()); err != nil {
+            log.Printf("trace: HTTPS server Shutdown: %v", err)
+        } else {
+            log.Printf("info: HTTPS Server shutdown successful!")
+        }
+        switch value {
+        case 0:
+            // Signal from signal_handler
+            log.Printf("info: Terminating...")
+            os.Exit(0)
+        }
     }
+}
 
-    log.Printf("info: https endpoint started")
-
-    // Listening should happen in a go-routine to prevent blocking
-    go func() {
-        log.Fatal(server.ListenAndServeTLS(*sw.ServerCertFlag, *sw.ServerKeyFlag))
-    }()
-
+func signal_handler(messenger chan<- int, wgroup *sync.WaitGroup) {
+    defer wgroup.Done()
     sigchannel := make(chan os.Signal, 1)
     signal.Notify(sigchannel,
         syscall.SIGTERM,
         syscall.SIGQUIT)
 
     <-sigchannel
-    log.Printf("info: Signal received. Shutting down...")
-    if err := server.Shutdown(context.Background()); err != nil {
-        log.Printf("trace: HTTP server Shutdown: %v", err)
-    } else {
-        log.Printf("info: Server shutdown successful!")
+    messenger <- 0
+    return
+}
+
+func monitor_certs(messenger chan<- int, wgroup *sync.WaitGroup) {
```

Comment:
I do not necessarily feel we need to compare the md5sum to find out whether the file has changed or not. Maybe we just need to check the modtime of the cert files. Using https://github.com/fsnotify/fsnotify

Reply:
I have used buit-in os.lstat in my latest change. Let's avoid using additional libraries.